include/seqan3/search/fm_index/bi_fm_index.hpp

```diff
@@ -201,36 +140,44 @@ class bi_fm_index
      *
      * No guarantee. \todo Ensure strong exception guarantee.
      */
+    template <std::ranges::Range text_t>
+        //!\cond
+        requires !is_collection
+        //!\endcond
     void construct(text_t const & text)
     {
-         // text must not be empty
+        static_assert(std::ranges::RandomAccessRange<text_t>, "The text must be a RandomAccessRange.");
+
+        // text must not be empty
         if (std::ranges::begin(text) == std::ranges::end(text))
             throw std::invalid_argument("The text that is indexed cannot be empty.");
 
-        this->text = &text;
-        if constexpr(is_collection)
-            rev_text = text | view::deep{std::view::reverse} | view::deep{view::persist} | std::view::reverse;
-        else
-            rev_text = std::view::reverse(text);
+        auto rev_text = std::view::reverse(text);
         fwd_fm.construct(text);
         rev_fm.construct(rev_text);
 
-        // does not work yet. segmentation fault in bi_fm_index_cursor snippet
-        // bi_fm_index tmp;
-        // tmp.text = &text;
-        // tmp.rev_text = std::view::reverse(*tmp.text);
-        // tmp.fwd_fm.construct(*tmp.text);
-        // tmp.rev_fm.construct(tmp.rev_text);
-        // std::swap(*this, tmp);
-        // this->text = &text;
-        // rev_text = std::view::reverse(text);
+        sigma = fwd_fm.sigma;
     }
 
     //!\overload
-    void construct(text_t &&) = delete;
+    template <std::ranges::Range text_t>
+        //!\cond
+        requires is_collection
+        //!\endcond
+    void construct(text_t const & text)
+    {
+        static_assert(std::ranges::RandomAccessRange<text_t>, "The text must be a RandomAccessRange.");
```

Comment:
```suggestion
        static_assert(std::ranges::BidirectionalRange<text_t>, "The text must be a BidirectionalRange.");
```

Also check the reference_t